src/mono/System.Private.CoreLib/src/System/Object.Mono.cs

```diff
@@ -13,5 +14,18 @@ public partial class Object
         [MethodImplAttribute(MethodImplOptions.InternalCall)]
         protected extern object MemberwiseClone();
 
+        [UnsafeAccessor(UnsafeAccessorKind.Method, Name = nameof(Finalize))]
+        private static extern void CallFinalize(object o);
+        private void GuardedFinalize()
+        {
+            try
+            {
+                CallFinalize(this);
+            }
+            catch (Exception ex) when (ExceptionHandling.IsHandledByGlobalHandler(ex))
```

Comment:
This is the key part of the change. 
The rest is just so that finalizer thread calls `object.GuardedFinalize()` instead of `object.Finalize()`